// @linaria/shaker/lib/dumpNode.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = dumpNode;

var _core = require("@babel/core");

const hooks = {
  Identifier: node => node.name,
  BinaryExpression: node => node.operator,
  NumericLiteral: node => node.value,
  StringLiteral: node => node.value
};

function isNode(obj) {
  return !!obj;
}

function dumpNode(node, alive = null, level = 0, idx = null) {
  let result = level === 0 ? '\n' : '';
  const prefix = level === 0 ? '' : `${'| '.repeat(level - 1)}${idx === null ? '|' : idx}${(idx || 0) < 10 ? '=' : ''}`;
  const {
    type
  } = node;
  result += `${prefix}${type}${type in hooks ? ` ${hooks[type](node)}` : ''}`;

  if (alive) {
    result += alive.has(node) ? ' ✅' : ' ❌';
  }

  result += '\n';
  const keys = _core.types.VISITOR_KEYS[type];

  for (const key of keys) {
    const subNode = node[key];
    result += `${'| '.repeat(level)}|-${key}\n`;

    if (Array.isArray(subNode)) {
      for (let i = 0; i < subNode.length; i++) {
        const child = subNode[i];
        if (child) result += dumpNode(child, alive, level + 2, i);
      }
    } else if (isNode(subNode)) {
      result += dumpNode(subNode, alive, level + 2);
    }
  }

  return result;
}
